cmd/bad-key-revoker/main.go

```diff
@@ -0,0 +1,360 @@
+				&unrevokedCert,
+				`SELECT cs.serial, c.registrationID, c.der FROM certificateStatus AS cs
+				JOIN certificates AS c
+				ON cs.serial = c.serial
+				WHERE cs.serial = ? AND cs.isExpired = false AND cs.status != ?`,
+				serial.CertSerial,
+				string(core.StatusRevoked),
+			)
+			if err != nil {
+				if db.IsNoRows(err) {
+					continue
+				}
+				return nil, err
+			}
+			unrevokedCert.RevokedBy = row.RevokedBy
+			unrevokedCerts = append(unrevokedCerts, unrevokedCert)
+		}
+		if len(batch) != bkr.certificatesBatchSize {
+			return unrevokedCerts, nil
+		}
+		initialID = batch[len(batch)-1].ID
+	}
+}
+
+func (bkr *badKeyRevoker) markRowChecked(row unchecked) error {
+	_, err := bkr.dbMap.Exec("UPDATE blockedKeys SET extantCertificatesChecked = true WHERE keyHash = ?", row.KeyHash)
+	return err
+}
+
+func (bkr *badKeyRevoker) resolveContacts(ids []int64) (map[int64]string, error) {
+	idToEmail := map[int64]string{}
+	for _, id := range ids {
+		var emails struct {
+			Contact []string
+		}
+		err := bkr.dbMap.SelectOne(&emails, "SELECT contact FROM registrations WHERE id = ?", id)
+		if err != nil {
+			return nil, err
+		}
+		if len(emails.Contact) == 0 {
+			continue
+		}
+		idToEmail[id] = strings.TrimPrefix(emails.Contact[0], "mailto:")
+	}
+	return idToEmail, nil
+}
+
+var emailTemplate = template.Must(template.New("email-template").Parse(`Hello,
+
+The public key associated with certificates which you have issued has been marked as compromised. As such we are required to revoke any certificates which contain this public key.
+
+The following currently unexpired certificates that you've issued contain this public key and have been revoked:
+{{range . -}}
+{{.}}
+{{end}}`))
+
+var emailSubject = "Certificates you've issued have been revoked due to key compromise"
+
+var maxSerials = 100
+
+func (bkr *badKeyRevoker) sendMessages(mapping map[string][]string) error {
+	err := bkr.mailer.Connect()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = bkr.mailer.Close()
+	}()
+	for email, serials := range mapping {
+		if len(serials) > maxSerials {
+			more := len(serials) - maxSerials
+			serials = serials[:maxSerials]
+			serials = append(serials, fmt.Sprintf("and %d more certificates.", more))
+		}
+		message := bytes.NewBuffer(nil)
+		err := emailTemplate.Execute(message, serials)
+		if err != nil {
+			return err
+		}
+		err = bkr.mailer.SendMail([]string{email}, emailSubject, message.String())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+var keyCompromiseCode = int64(revocation.KeyCompromise)
+
+func (bkr *badKeyRevoker) invoke() (bool, error) {
+	// find rows to work on
+	uncheckedRows, err := bkr.selectUncheckedRows()
+	if err != nil {
+		return false, err
+	}
+	if len(uncheckedRows) == 0 {
+		return true, nil
+	}
+	for _, row := range uncheckedRows {
+		// find everything in keyHashToSerial with certificateStatus.status != revoked
+		unrevokedCerts, err := bkr.findUnrevoked(row)
```

Comment:
I don't really find this convincing, for a handful of reasons, the first of which is it significantly reduces the impact of this tooling. The tool would end up just taking care of the 'easy' quick to fix problems, while offloading the complex large tasks to SRE, something we are explicitly trying to fix with the tool.

Secondly it would just mask the real problem, _most of the time_ we are going to see only a handful of certificates with the same SPKI, and would only trigger the safeguard when there were lots of certificates using the same SPKI, meaning the bad behavior would likely go on for a long time before it was caught. I think this is more appropriately prevented by proper unit, integration, and staging testing and monitoring the number of things we are revoking via the daemon.

Reply:
We may be thinking about different "bad behavior." I'm thinking of something that e.g. triggers revocation for all certificates due to a missed or incorrect WHERE clause or some other logic error. This would be a terrible problem, because certificates cannot be unrevoked. Because it's such a terrible problem it deserves additional belt-and-suspenders protection.

> The tool would end up just taking care of the 'easy' quick to fix problems, while offloading the complex large tasks to SRE, something we are explicitly trying to fix with the tool.

I'm not proposing to fall back to manual revocation if the list exceeds the configured maximum. I'm proposing that we have essentially a circuit breaker - if the tool hits the configured maximum in one go, someone has to check that it's alright, and if it is, set the maximum higher. This is still much less work than manual revocation.

Reply:
> We may be thinking about different "bad behavior." I'm thinking of something that e.g. triggers revocation for all certificates due to a missed or incorrect WHERE clause or some other logic error. This would be a terrible problem, because certificates cannot be unrevoked. Because it's such a terrible problem it deserves additional belt-and-suspenders protection.

I'm still not really convinced this would be the best way to catch this kind of behavior. Something that selects _everything_ seems unlikely to be randomly triggered, it'll either always happen, and get caught by testing, or not.

> I'm not proposing to fall back to manual revocation if the list exceeds the configured maximum. I'm proposing that we have essentially a circuit breaker - if the tool hits the configured maximum in one go, someone has to check that it's alright, and if it is, set the maximum higher. This is still much less work than manual revocation.

This is still not a simple process, it requires a config change, review, config deploy, component restart, config revert, review, and config deploy given the current SRE process.

Reply:
> I'm still not really convinced this would be the best way to catch this kind of behavior. Something that selects everything seems unlikely to be randomly triggered, it'll either always happen, and get caught by testing, or not.

Fair enough, but if the consequences of being wrong are "every site with a Let's Encrypt certificate has an outage," I'm not content with "unlikely," I want additional guarantees.